File: sample-conference-videochat/sample-conference-videochat/Components/Core/UserPorfile/QBProfile.m
//
//  QBProfile.m
//  LoginComponent
//
//  Created by Andrey Ivanov on 02/06/16.
//  Copyright © 2016 Quickblox. All rights reserved.
//

#import "QBProfile.h"
#import <Quickblox/Quickblox.h>

NSString *const kQBProfile = @"curentProfile";
NSString *const kQBUser = @"qbuser";

@interface QBProfile()

@property (strong, nonatomic, readwrite) QBUUser *userData;

@end

@implementation QBProfile

+ (instancetype)currentProfile {
    
    return [[self alloc] init];
}

- (instancetype)init {
    
    self = [super init];
    if (self) {
        
        [self loadProfile];
    }
    
    return self;
}

- (OSStatus)synchronize {
    
    NSParameterAssert(self.userData);
    return [self saveData:self forKey:kQBProfile];
}

- (OSStatus)synchronizeWithUserData:(QBUUser *)userData {
    
    self.userData = userData;
    OSStatus status = [self synchronize];
    
    return status;
}

- (void)loadProfile {
    
    QBProfile *profile = [self loadObjectForKey:kQBProfile];
    self.userData = profile.userData;
}

- (OSStatus)clearProfile {
    
    OSStatus success = [self deleteObjectForKey:kQBProfile];
    
    self.userData = nil;
    
    return success;
}

#pragma mark - NSCoding

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    
    self = [super init];
    
    if (self) {
        
        _userData = [aDecoder decodeObjectForKey:kQBUser];
    }
    
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    
    [aCoder encodeObject:self.userData forKey:kQBUser];
}

- (OSStatus)saveData:(id)data forKey:(NSString *)key {
    
    NSMutableDictionary *keychainQuery = [self getKeychainQueryForKey:key];
    SecItemDelete((__bridge CFDictionaryRef)keychainQuery);
    keychainQuery[(__bridge id) kSecValueData] = [NSKeyedArchiver archivedDataWithRootObject:data];
    
    return SecItemAdd((__bridge CFDictionaryRef)keychainQuery, NULL);
}

#pragma mark - keychain

- (id)loadObjectForKey:(NSString *)key {
    
    id ret = nil;
    
    NSMutableDictionary *keychainQuery = [self getKeychainQueryForKey:key];
    
    keychainQuery[(__bridge id) kSecReturnData] = (id) kCFBooleanTrue;
    keychainQuery[(__bridge id) kSecMatchLimit] = (__bridge id) kSecMatchLimitOne;
    CFDataRef keyData = NULL;
    
    if (SecItemCopyMatching((__bridge CFDictionaryRef)keychainQuery, (CFTypeRef *)&keyData) == noErr) {
        
        ret = [NSKeyedUnarchiver unarchiveObjectWithData:(__bridge NSData *)keyData];
    }
    
    if (keyData) CFRelease(keyData);
    
    return ret;
}

- (OSStatus)deleteObjectForKey:(NSString *)key {
    
    NSMutableDictionary *keychainQuery = [self getKeychainQueryForKey:key];
    
    return SecItemDelete((__bridge CFDictionaryRef)keychainQuery);
}

- (NSMutableDictionary *)getKeychainQueryForKey:(NSString *)key {
    
    return [@{(__bridge id) kSecClass : (__bridge id) kSecClassGenericPassword,
              (__bridge id) kSecAttrService : key,
              (__bridge id) kSecAttrAccount : key,
              (__bridge id) kSecAttrAccessible : (__bridge id) kSecAttrAccessibleAfterFirstUnlock} mutableCopy];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Core/UserPorfile/QBProfile.swift
//
//  QBProfile.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import Foundation
import Quickblox
import Security

struct QBProfileConstants {
    static let kQBProfile = "curentProfile"
    static let kQBUser = "qbuser"
}

struct QBProfileSecConstants {
    static let kSecClassValue = NSString(format: kSecClass)
    static let kSecAttrAccountValue = NSString(format: kSecAttrAccount)
    static let kSecValueDataValue = NSString(format: kSecValueData)
    static let kSecClassGenericPasswordValue = NSString(format: kSecClassGenericPassword)
    static let kSecAttrServiceValue = NSString(format: kSecAttrService)
    static let kSecAttrAccessibleValue = NSString(format: kSecAttrAccessible)
    static let kSecMatchLimitValue = NSString(format: kSecMatchLimit)
    static let kSecReturnDataValue = NSString(format: kSecReturnData)
    static let kSecMatchLimitOneValue = NSString(format: kSecMatchLimitOne)
    static let kSecAttrAccessibleAfterFirstUnlockValue = NSString(format: kSecAttrAccessibleAfterFirstUnlock)
}

class QBProfile: NSObject, NSCoding, NSSecureCoding  {
    
    static var supportsSecureCoding: Bool {
        return true
    }
    
    var userData: QBUUser?
    
    init(userData: QBUUser?) {
        super.init()
        self.userData = userData
    }
    
    override init() {
        super.init()
        loadProfile()
    }
    
    /**
     *  Synchronize current profile in keychain.
     *
     *  @return whether synchronize was successful
     */
    func synchronize() -> OSStatus {
        
        assert(self.userData != nil, "Invalid parameter not satisfying: userData != nil")
        debugPrint("self.userData \(String(describing: self.userData))")
        return self.saveData(self.userData as Any, forKey: QBProfileConstants.kQBProfile)
    }
    
    convenience required init?(coder aDecoder: NSCoder) {
       let userData = aDecoder.decodeObject(forKey: QBProfileConstants.kQBProfile) as? QBUUser
        self.init(userData: userData)
    }

    func encode(with aCoder: NSCoder) {
        aCoder.encode(userData, forKey: QBProfileConstants.kQBProfile)
    }

    /**
     *  Synchronize user data in keychain.
     *
     *  @param userData user data to synchronize
     *
     *  @return whether synchronize was successful
     */
    public func synchronizeWithUserData(userData: QBUUser) -> OSStatus {
        self.userData = userData;
        let status: OSStatus = synchronize()
        return status;
    }
    
    private func loadProfile() {
        if let profile = self.loadObject(forKey: QBProfileConstants.kQBProfile) {
            self.userData = profile;
            debugPrint("self.userData \(String(describing: self.userData))")
        } else {
            return
        }
    }
    
    /**
     *  Remove all user data.
     *
     *  @return Whether clear was successful
     */
    public func clearProfile() -> OSStatus {
        let success: OSStatus = self.deleteObjectForKey(key: QBProfileConstants.kQBProfile)
        self.userData = nil;
        return success;
    }
    
    func saveData(_ data: Any?, forKey key: String?) -> OSStatus {
        
        var keychainQuery = getKeychainQueryFor(key: key!)
        SecItemDelete((keychainQuery as CFDictionary?)!)
        if let dataUser = data {
            keychainQuery?[QBProfileSecConstants.kSecValueDataValue] = NSKeyedArchiver.archivedData(withRootObject: dataUser)
        }
        return SecItemAdd(keychainQuery! as CFDictionary, nil)
    }
    
    // MARK: - Keychain
    func loadObject(forKey key: String?) -> QBUUser? {
        
        var ret: QBUUser? = nil
        
        var keychainQuery = getKeychainQueryFor(key: key!)
        if let aTrue = kCFBooleanTrue {
            keychainQuery?[QBProfileSecConstants.kSecReturnDataValue] = aTrue
        }
        keychainQuery?[QBProfileSecConstants.kSecMatchLimitValue] = QBProfileSecConstants.kSecMatchLimitOneValue
        var keyData: AnyObject?
        let status = SecItemCopyMatching(keychainQuery! as CFDictionary, &keyData)
        if status == noErr {
            if let aData = keyData as! Data? {
                ret = NSKeyedUnarchiver.unarchiveObject(with: aData) as? QBUUser
              return ret
            }
        }
        return ret
    }
    
    private func deleteObjectForKey(key: String) -> OSStatus {
        
        let keychainQuery = self.getKeychainQueryFor(key: key)
        return SecItemDelete(keychainQuery! as CFDictionary)
    }
    
    private func getKeychainQueryFor(key: String) -> [AnyHashable : Any]? {
        
            return ([QBProfileSecConstants.kSecClassValue: QBProfileSecConstants.kSecClassGenericPasswordValue, QBProfileSecConstants.kSecAttrServiceValue: key, QBProfileSecConstants.kSecAttrAccountValue: key, QBProfileSecConstants.kSecAttrAccessibleValue: QBProfileSecConstants.kSecAttrAccessibleAfterFirstUnlockValue])
    }
}
